docs: Add the section for how to make a PR. (#94)
openedx/docs.openedx.org: source/developers/quickstarts/setup_a_backend_dev_environment.rst

Quick Start: First Open edX Pull Request
########################################

.. contents:: Steps to Making your First Pull Request
   :local:
   :class: no-bullets

Welcome to the Quickstart guide for new Open edX contributors. By the end of
this tutorial, you will have completed all the steps necessary to begin
contributing to the Open edX project.

These include:

* Setting up your development environment;
* Making your first commit;
* Submitting your first pull request (PR);
* Signing your Contributor License Agreement (CLA).

We assume you are comfortable with the command line, and understand the basics
of using Git, GitHub, and python basics.

This Quickstart is written for first-time contributors to the Open edX project.
If you are a plug-in developer or a frontend developer, please see our
Quickstart guide for plug-in developers <link coming soon> or our Quickstart for
frontend developers <link coming soon>.

System Requirements
*******************

For the smoothest experience, we recommend that your computer has at least 16 GB
of RAM, 2 CPUs, and at least 50 GB of Free disk space.

Before you start
****************

> This tutorial is written for users of Mac OS, either with an Intel or an Arm

Reviewer: macOS

> (Apple Silicone, M1, M2, etc.) processor. Additionally, you will need to have the following

Reviewer 2: ```suggestion
(Apple Silicon, M1, M2, etc.) processor. Additionally, you will need to have the following
```

installed or configured, and know at least the basics of using them, before
proceeding:

* Git (see `GitHub's set up Git guide
  <https://help.github.com/en/github/getting-started-with-github/set-up-git>`_)

* A `GitHub account <https://github.com/signup>`_
  * Additionally, we strongly recommend setting up 2-factor authentication

* `Python 3.8 <https://www.python.org/downloads/>`_

* Your favorite code editor (our team uses `VSCode
  <https://code.visualstudio.com/download>`_, `Emacs
  <https://emacsformacosx.com/>`_ or via `homebrew`_, `Vim
  <https://github.com/macvim-dev/macvim>`_ or via `homebrew`_, and others)

.. _homebrew: https://brew.sh

Install Docker
**************

> Install `Docker for Mac <https://docs.docker.com/docker-for-mac/>`_ and launch

Reviewer: macOS

it. You can check that it is running correctly with::

    docker run --rm hello-world

You should see a message that starts with the following::

   Hello from Docker!
   This message shows that your installation appears to be working correctly.

On macOS, by default, Docker allocates at most 2 GB of RAM to containers; the
Open edX software requires at least 8 GB. You should follow these instructions
from the `official Docker documentation
<https://docs.docker.com/docker-for-mac/#advanced>`_ to allocate at least 8 GB
to the Docker daemon.

Installing Tutor Nightly
************************

Navigate into a folder to hold your Open edX repositories.
We will assume you're using ``~/openedx``, but you can choose any folder.

.. code-block:: bash

  mkdir -p ~/openedx
  cd ~/openedx

Create a Python 3.8 virtual environment (we'll call it ``tutor-venv``) and activate it:

.. code-block:: bash

  python3.8 -m venv tutor-venv
  . tutor-venv/bin/activate

> Install Tutor. This is a tool that helps you run Open edX.

Reviewer 3: ```suggestion
Install Tutor. This is a tool that helps you run the Open edX project.
```

We will install the "Nightly" version of Tutor, which means it will run the latest

> version of Open edX (as opposed to the stable version of Open edX):

Reviewer 3: ```suggestion
version of the code (as opposed to the most recent named release):
```

Reviewer 3: Specifically suggesting this because I think people would better understand "named release" over "stable version"

.. code-block:: bash

  git clone --branch=nightly https://github.com/overhangio/tutor.git
  pip install -e "./tutor[full]"

> If you are using ARM-64, then install this extra plugin and enable it

Reviewer 2: ```suggestion
If you are using ARM-64 (such as an M1 or M2 Mac), then install this extra plugin and enable it
```

(if you're not, then skip this step):

.. code-block:: bash

  pip install git+https://github.com/open-craft/tutor-contrib-arm64
  tutor plugins enable arm64
  tutor config save

Finally, let's configure and provision your Open edX instance!
You will be asked a couple questions.
Answer them however you like, although the default answers will work fine.

.. code-block:: bash

  tutor dev quickstart

Depending on your system and your Internet connection speed,
this could take anywhere from five minutes to over an hour,
so go get a coffee and come back for the next part.

Working with a fork
*******************

At this point you should have a Tutor installation that is suitable for
development, but you're still missing a practical way to edit the code, test
it locally, and then contribute it back.

For the purposes of this tutorial, you'll be modifying code in the

> ``_edx-platform`` repository, where the Open edX backend code lives.  Let's

Reviewer 3: ```suggestion
``edx-platform`` repository, where the Open edX backend code lives.  Let's
```

start by creating your own personal "fork" of it. A "fork" is essentially your
own copy of the repository. `See here <https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_ to learn more about forks.

**Forking edx-platform**

> Assuming you're logged in to Github, forking a repository is easy.  Visit the

Reviewer 3: ```suggestion
Assuming you're logged in to GitHub, forking a repository is easy.  Visit the
```

``edx-platform`` repository at this URL:

https://github.com/openedx/edx-platform

Now, click the :guilabel:`Fork` button on the top right, and in the next
screen, select your personal account as the owner.  After you click the
:guilabel:`Create fork` button, you'll be taken to your own version of the
``edx-platform`` repository.

**Cloning your fork**

> Your ``edx-platform`` currently only exists in the Github servers.  You'll now

Reviewer 3: ```suggestion
Your ``edx-platform`` currently only exists on the GitHub servers.  You'll now
```

Quick Start: First Open edX Pull Request
########################################

.. contents:: Steps to Making your First Pull Request
   :local:
   :class: no-bullets

Welcome to the Quickstart guide for new Open edX contributors. By the end of
this tutorial, you will have completed all the steps necessary to begin
contributing to the Open edX project.

These include:

* Setting up your development environment;
* Making your first commit;
* Submitting your first pull request (PR);
* Signing your Contributor License Agreement (CLA).

We assume you are comfortable with the command line, and understand the basics
of using Git, GitHub, and python basics.

This Quickstart is written for first-time contributors to the Open edX project.
If you are a plug-in developer or a frontend developer, please see our
Quickstart guide for plug-in developers <link coming soon> or our Quickstart for

> frontend developers <link coming soon>.

Reviewer: Adding links to these pages

``edx-platform`` repository.

**Cloning your fork**

Your ``edx-platform`` currently only exists in the Github servers.  You'll now
create a local copy of it (a "clone").

First, fetch the git URL of your fork.  Navigate to its web page (to which you
were taken after creation), click on the :guilabel:`Code` button, select
the **HTTPS** tab, and copy the URL given.  It should look like this:

``https://github.com:<your_github_username>/edx-platform.git``

Now, from the same top level directory you created above, clone the repository
as follows:

.. code-block:: bash

   cd ~/openedx
   git clone  https://github.com:<your_github_username>/edx-platform.git

You'll now have an ``edx-platform`` directory containing a local clone of your
fork.  It is not yet wired into your Tutor development environment, though.
This is what you'll do next.

**Mounting edx-platform**

To have Tutor run your local fork of edx-platform, you have to tell it to do so
on start up.  It is a simple CLI parameter that points Tutor to the directory where
the code lives.

As a first step, fire up a one-off LMS container while mounting your local
checkout:

.. code-block:: bash

>    tutor dev run --mount=${PWD}/edx-platform lms bash

Reviewer 3: Is it clear what PWD means? (just thinking from the perspective of someone hitting a lot of this for the first time, things can be kind of confusing when you're jamming in a bunch of new knowledge) cc @arbrandes

Reviewer 4: I guess this is part of the discussions about assumptions.  `$PWD` is reasonably basic command-line stuff.

Reviewer 3: Ya, that's why I asked. This one feels squishy but I guess I have no strong thoughts.

Author: I think it's fine to leave, they're trusting us to run a lot of commands already.

.. image:: /_images/developers_quickstart_first_pr/learner_dashboard_after.png
   :alt: Learner Dashboard page after we add the welcome message.

Make a Commit
=============

Now that you've saved your changes, you can make a commit. First make a new
branch with the name ``developer_quickstart``::

    git checkout -b developer_quickstart

Then we can create the actual commit. Note that Open edX commit messages should
follow our `conventional commit <https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html>`_
practices. In our case, we're making a new feature, so our commit message must
be prefixed with "feat:" like so::

    commit -a -m "feat: add welcome message to learner dashboard"

Push the Commit to Your Fork
============================

Now push your changes to a new branch in your fork::

    git push --set-upstream origin developer_quickstart

Create A Pull Request
*********************

Branch Name: developer_quickstart

# Go to your fork.

``https://github.com/<your_github_username>/edx-platform``

At the top of the page you'll see a section that will suggest that you make a

> new pull request.  Go ahead an click the big green button.

Reviewer 3: ```suggestion
new pull request.  Go ahead and click the big green button.
```

Setup a BackEnd Dev Environment
###############################

Branch Name: <github_username>/quickstart

# Go to your fork.

``https://github.com/<your_github_username>/edx-platform``

At the top of the page you'll see a section that will suggest that you make a

> new pull request.  Go ahead an click the big green button.

Reviewer 5: "and"
